imagepipeline/src/main/java/com/facebook/imagepipeline/producers/HttpUrlConnectionNetworkFetcher.java

```diff
@@ -47,37 +54,33 @@ public void fetch(final FetchState fetchState, final Callback callback) {
         new Runnable() {
           @Override
           public void run() {
-            HttpURLConnection connection = null;
-            Uri uri = fetchState.getUri();
-            String scheme = uri.getScheme();
-            String uriString = fetchState.getUri().toString();
-            while (true) {
-              String nextUriString;
-              String nextScheme;
+              HttpURLConnection connection = null;
+              Uri uri = fetchState.getUri();
+              String uriString = uri.toString();
               InputStream is;
               try {
-                URL url = new URL(uriString);
-                connection = (HttpURLConnection) url.openConnection();
-                nextUriString = connection.getHeaderField("Location");
-                nextScheme = (nextUriString == null) ? null : Uri.parse(nextUriString).getScheme();
-                if (nextUriString == null || nextScheme.equals(scheme)) {
+                  connection = createGetConnection(uriString);
+                  int redirectCount = 0;
+                  while (connection.getResponseCode() / 100 == 3 && redirectCount < MAX_REDIRECT_COUNT) {
```

Comment:
```
while (connection.getResponseCode() / 100 == 3 &&
    ++redirectCount <= MAX_REDIRECT_COUNT) {
  ...
}
```

Notice the preincrement in the check **after**  the status code check. Now after the loop one of those two things is true:
* redirectCount <= MAX_REDIRECT_COUNT  && code != 3xx
* redirectCount > MAX_REDIRECT_COUNT && code == 3xx